### PR TITLE
Do not rely on Referer HTTP header data when creating topic on Discourse

### LINKED ISSUE #736: Do not use referer to identify destination url on topic creation on discourse
First of all, read this : https://blog.mozilla.org/security/2018/01/31/preventing-data-leaks-by-stripping-path-information-in-http-referrers/

Modern browsers offers an option on privacy settings that remove any url content in referer field, in order to prevent personal data leak.

In consequence, this code does not contains the good URL, and then, the topic does not links to the good document : 

https://github.com/c2corg/v6_api/blob/3a5d5202da0ee49903b7ce495e141c41dce27892/c2corg_api/views/forum.py#L91-L94

We must change the way we get the destination url : 

* rebuild it from document_id, lang, and title (we have everything we need, except the disctinction between prod/demo domain. We could hard-write production domain 
* or send it in the body request

### PR DESCRIPTION
**Pull request by tristan0x**
Assume that the web browser wiped the referrer data (path, query, ...)
because calling the API endpoint.
This changes rebuilds the URL manually by using both the document identifier
passed in the request payload and the database to retrieve the document type.

Fixes #736

### CODE REVIEW

**Comment by tristan0x** [comment]
ping @cbeauchesne

File: c2corg_api/views/forum.py [hunk 1]
```diff
@@ -86,12 +97,25 @@ def __init__(self, request):
     def collection_post(self):
         settings = self.request.registry.settings
 
+        import pprint
         locale = self.request.validated['locale']
+        document = self.request.validated['document']
+        log.warning(pprint.pformat(dir(locale)))
```
Comment (cbeauchesne) [inline c2corg_api/views/forum.py:103]: Is warning the good level? Maybe debug ?

File: c2corg_api/views/forum.py [hunk 2]
```diff
@@ -86,12 +97,25 @@ def __init__(self, request):
...
+        document_type = association_keys_for_types[document.type]
+
+        referer = self.request.referer or '/'
+        referer_path = parse_url(referer).path
+        # only keep the left part of referer path
+        referer_scheme_host = referer.rsplit(referer_path, 1)[0]
```
Comment (cbeauchesne) [inline c2corg_api/views/forum.py:109]: if `self.request.referer` is missing, then `referer_scheme_host` will be an empty string.

As the host site will always be www.camtocamp.org, I think we should rather keep it very simple: 

``` python
content = '<a href="https://www.camptocamp.org{}">{}</a>'.format(referer_path)
``` 

And IMO,  rename `referer_path` to `document_path`

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

File: c2corg_api/views/forum.py (hunk 1 above)
Comment (tristan0x) [reply c2corg_api/views/forum.py]: This was debug code I removed in 7bf9533

File: c2corg_api/views/forum.py (hunk 2 above)
Comment (tristan0x) [reply c2corg_api/views/forum.py]: I applied your suggestion in 7bf9533

**Comment by tristan0x** [comment]
One test fails in Travis CI fails with 7bf9533

> c2corg_api/tests/views/test_association.py:454: AssertionError

So far, I cannot reproduce the issue on my workstation. I will try later on with a fresh database.

**Comment by lbesson** [comment]
could you rebase against latest master?

**Comment by tristan0x** [comment]
> could you rebase against latest master?

It's done @lbesson 
Sorry for the delay...

**Comment by lbesson** [comment]
No problem, I'm just trying to make teh api live again.
Your response time is fine to me :)

I didn't follow this from the beginning, what is the status in your opinion? Should it be merged? Is som additional work required?

**Comment by tristan0x** [comment]
> No problem, I'm just trying to make the api project live again.
> Your response time is fine to me :)
> 
> I didn't follow this from the beginning, what is the status in your opinion? Should it be merged? Is some additional work required?

Charles's points have been addressed, so it is mergeable in my opinion. Maybe @cbeauchesne can take a last look before we merge this.
Cheers!

**Comment by cbeauchesne** [comment]
If I remember correctly, it's all good.!

**Comment by lbesson** [comment]
Great. Let's merge then.

**Pull request merged**